### PR TITLE
key_rev_value-based revision culling

### PR DESCRIPTION
**Pull request by eevans**
Moves title iteration out to its own module, and re-uses it to create a Parsoid-specific thinning script.  This new script performs corresponding deletes to data-parsoid as part of a batch operation.  Since both entries share the same partition key, this is atomic, isolated, and performant.

This is still quite rough, but I wanted to get it put up, and more eyes on it, before taking it much further.

Comments appreciated.

### CODE REVIEW

File: maintenance/thin_out_key_rev_value_data.js [hunk 1]
```diff
@@ -169,78 +144,4 @@ function processRow (row) {
...
-        .catch(function(e) {
-            console.log(res.pageState);
-            console.log(e);
-            throw e;
-        });
-    });
-}
-
-return processRows(startOffset.pageState);
+return iterateTable(client, startOffset, processRow);
```
Comment (Pchelolo) [inline maintenance/thin_out_key_rev_value_data.js]: Where's the `tableName` parameter?

File: maintenance/thin_out_parsoid.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    process.exit(1);
+}
+
+function makeClient() {
+    return new cassandra.Client({
+        contactPoints: [argv.host],
+        authProvider: new cassandra.auth.PlainTextAuthProvider('cassandra', 'cassandra'),
+        socketOptions: { connectTimeout: 10000 },
+    });
+}
```
Comment (Pchelolo) [inline maintenance/thin_out_parsoid.js]: Could be nice to move out this one too and reuse

File: maintenance/thin_out_parsoid.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    if ((total % 500000) === 0) {
+        console.log(new Date() + ': processed ' + total + ' total entries');
+    }
+
+    // Thin-out
+    if ((counts.rev > 0 && counts.render > 0)
+        || (counts.rev === 0 && counts.render > 0
+            // Enforce a grace_ttl of 86400
+            && (Date.now() - row.tid.getDate()) > 86400000)
+        || (counts.rev > 0 && row.tid.getDate() <  Date.parse('2015-12-31T23:59-0000'))) {
```
Comment (Pchelolo) [inline maintenance/thin_out_parsoid.js]: Should that date be a parameter?

File: maintenance/thin_out_parsoid.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        console.log('-- Deleting:', row._token.toString(), row.tid.getDate().toISOString(), keys.rev);
+
+        var delQuery1 = 'DELETE FROM ' + htmlTable + ' WHERE "_domain" = ? AND key = ? AND rev = ? AND tid = ?';
+        var delQuery2 = 'DELETE FROM ' + dataTable + ' WHERE "_domain" = ? AND key = ? AND rev = ? AND tid = ?';
```
Comment (Pchelolo) [inline maintenance/thin_out_parsoid.js]: Should we delete from the `section.offsets` table too?

Comment (gwicke) [inline maintenance/thin_out_parsoid.js]: +1

File: maintenance/thin_out_key_rev_value_data.js (hunk 1 above)
Comment (eevans) [inline maintenance/thin_out_key_rev_value_data.js]: Yup, good catch.

File: maintenance/thin_out_parsoid.js (hunk 2 above)
Comment (eevans) [inline maintenance/thin_out_parsoid.js]: Ok.

File: maintenance/thin_out_parsoid.js (hunk 4 above)
Comment (eevans) [inline maintenance/thin_out_parsoid.js]: Done.

**Comment by eevans** [comment]
This seems to have implicitly passed review, and since it is now being used on WMF production systems, it's probably worth having master up-to-date with the status quo.

**Pull request merged**